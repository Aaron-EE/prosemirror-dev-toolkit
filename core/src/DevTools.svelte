<script lang="ts">
  import { onMount } from 'svelte'
  import type { EditorView } from 'prosemirror-view'
  import { setContext } from '$context'

  import FloatingBtn from './FloatingBtn.svelte'
  import FloatingDock from './FloatingDock.svelte'

  import type { ButtonPosition } from './types'

  export let view: EditorView,
    devToolsExpanded = false,
    buttonPosition: ButtonPosition = 'bottom-right'

  setContext('editor-view', {
    view,
  })

  onMount(() => {
    const html = document && document.querySelector('html')
    if (devToolsExpanded && html) {
      html.style.marginBottom = '341px'
    }
  })

  function handleFloatingBtnClick() {
    devToolsExpanded = true
    const html = document && document.querySelector('html')
    if (html) {
      html.style.marginBottom = '341px'
    }
  }
  function handleFloatingDockClose() {
    devToolsExpanded = false
    const html = document && document.querySelector('html')
    if (html) {
      html.style.marginBottom = null
    }
  }
</script>

<section>
  {#if devToolsExpanded}
    <FloatingDock onClose={handleFloatingDockClose} />
  {:else}
    <FloatingBtn on:click={handleFloatingBtnClick} {buttonPosition} />
  {/if}
</section>

<style lang="scss">
  * {
    --font-sans: Helvetica Neue, Calibri Light, Roboto, sans-serif;
    --font-small: 11px;
    --font-medium: 13px;
    --font-large: 16px;
    --height-tabs-menu: 48px;
    --color-black: #222;
    --color-blue-bg: #363755;
    --color-blue-light: #85d9ef;
    /* --color-green-text: #b8e248; */
    --color-gray-light: #d3d3d9;
    --color-green: green; // diff inserted
    --color-green-light: #87cc86; // diff deleted color
    --color-purple-dark: rgb(80, 68, 93); // list darker bg
    --color-purple: rgb(96, 76, 104); // borders in eg lists
    --color-purple-light: rgb(99, 99, 123); // list text color
    --color-red: #d66363; // diff deleted
    --color-red-gray: rgb(187, 145, 163); // h2 color
    --color-red-light: rgb(255, 162, 177);
    --color-yellow: #eaea37; // diff updated
    --color-white: #fff;
  }
  :root {
    --tree-view-base00: #363755;
    --tree-view-base01: #604d49;
    --tree-view-base02: #6d5a55;
    --tree-view-base03: #d1929b;
    --tree-view-base04: #b79f8d;
    --tree-view-base05: #f9f8f2;
    --tree-view-base06: #f7f4f1;
    --tree-view-base07: #faf8f5;
    --tree-view-base08: #fa3e7e;
    --tree-view-base09: #fd993c;
    --tree-view-base0A: #f6bf81;
    --tree-view-base0B: #b8e248;
    --tree-view-base0C: #b4efe4;
    --tree-view-base0D: #85d9ef;
    --tree-view-base0E: #be87ff;
    --tree-view-base0F: #d6724c;
  }
  section {
    font-family: var(--font-sans);
    font-size: var(--font-medium);
  }
</style>
