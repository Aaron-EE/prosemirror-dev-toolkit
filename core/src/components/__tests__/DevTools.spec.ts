/**
 * @jest-environment jsdom
 */

import { render, fireEvent } from '@testing-library/svelte'
import { EditorView } from 'prosemirror-view'
import { vi } from 'vitest'

import DevTools from '../DevTools.svelte'
import { createEditorInstance } from './pm-test-utils'

let editorView: EditorView

describe('DevTools.svelte', () => {
  beforeAll(() => {
    const el = document.createElement('div')
    document.body.appendChild(el)
    el.id = 'pm-editor'
    editorView = createEditorInstance(el)

    vi.stubGlobal('prompt', (str: string) => undefined)
  })

  it('should render', () => {
    const results = render(DevTools, {
      props: {
        view: editorView,
        buttonPosition: 'top-left',
        devToolsExpanded: false
      }
    })

    expect(results.container).toBeInTheDocument()
    expect(results.container).toMatchSnapshot()
  })

  it('should open by clicking the floating button', async () => {
    const results = render(DevTools, { props: { view: editorView, devToolsExpanded: true } })
    const button = results.container.querySelector('button')
    expect(button).not.toBeNull()

    await fireEvent.click(button as HTMLElement)
    const snapshotsTabBtn = await results.findByText('SNAPSHOTS')
    await fireEvent.click(snapshotsTabBtn as HTMLElement)

    expect(results.container).toBeInTheDocument()
    expect(results.container).toMatchSnapshot()
  })
})
