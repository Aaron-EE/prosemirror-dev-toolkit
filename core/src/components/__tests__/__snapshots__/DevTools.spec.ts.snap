// Vitest Snapshot v1

exports[`DevTools.svelte > should open by clicking the floating button 1`] = `
<body>
  <div
    class="ProseMirror"
    contenteditable="true"
    id="pm-editor"
    translate="no"
  >
    <p>
      <br
        class="ProseMirror-trailingBreak"
      />
    </p>
  </div>
  <div>
    <section
      class="dev-tools s-eHt3GWKXLdZl"
    >
      <div
        class="floating-dock-wrapper s-g5EO2O14MaeA"
      >
        <div
          class="floating-dock s-g5EO2O14MaeA"
          style="top: 50%; height: 50%;"
        >
          <div
            class="resizing-div s-g5EO2O14MaeA"
          />
           
          <div
            class="container s-g5EO2O14MaeA"
          >
            <div
              class="s-g5EO2O14MaeA"
            >
              <button
                class="snap-save-btn s-g5EO2O14MaeA"
              >
                Save snapshot
              </button>
               
              <button
                class="snap-import-btn s-g5EO2O14MaeA"
              >
                Import snapshot
              </button>
               
              <button
                aria-label="Close dev-toolkit button"
                class="close-btn s-g5EO2O14MaeA"
              >
                X
              </button>
            </div>
             
            <input
              accept=".json"
              class="s-g5EO2O14MaeA"
              multiple=""
              style="display: none;"
              type="file"
            />
             
            <ul
              class="tabs-menu s-G9CdWS8GObkb"
            >
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  STATE
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  HISTORY
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  PLUGINS
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  SCHEMA
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  STRUCTURE
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb active"
                >
                  SNAPSHOTS
                </button>
              </li>
            </ul>
            <!--&lt;TabsMenu&gt;-->
             
            <section
              class="split-view s-f-Fay03D4qzv"
            >
               
              <div
                class="right-panel s-qFD0hR3IuoME"
                slot="right"
              >
                <div
                  class="no-snapshots s-qFD0hR3IuoME"
                >
                  Save snapshots by clicking "Save snapshot" button.
                </div>
              </div>
            </section>
            <!--&lt;SplitView&gt;-->
            <!--&lt;SnapshotsTab&gt;-->
          </div>
        </div>
      </div>
      <!--&lt;FloatingDock&gt;-->
    </section>
    <!--&lt;DevTools&gt;-->
  </div>
</body>
`;

exports[`DevTools.svelte > should render 1`] = `
<body>
  <div
    class="ProseMirror"
    contenteditable="true"
    id="pm-editor"
    translate="no"
  >
    <p>
      <br
        class="ProseMirror-trailingBreak"
      />
    </p>
  </div>
  <div>
    <section
      class="dev-tools s-eHt3GWKXLdZl"
    >
      <div
        class="floating-dock-wrapper s-g5EO2O14MaeA"
      >
        <div
          class="floating-dock s-g5EO2O14MaeA"
          style="top: 50%; height: 50%;"
        >
          <div
            class="resizing-div s-g5EO2O14MaeA"
          />
           
          <div
            class="container s-g5EO2O14MaeA"
          >
            <div
              class="s-g5EO2O14MaeA"
            >
              <button
                class="snap-save-btn s-g5EO2O14MaeA"
              >
                Save snapshot
              </button>
               
              <button
                class="snap-import-btn s-g5EO2O14MaeA"
              >
                Import snapshot
              </button>
               
              <button
                aria-label="Close dev-toolkit button"
                class="close-btn s-g5EO2O14MaeA"
              >
                X
              </button>
            </div>
             
            <input
              accept=".json"
              class="s-g5EO2O14MaeA"
              multiple=""
              style="display: none;"
              type="file"
            />
             
            <ul
              class="tabs-menu s-G9CdWS8GObkb"
            >
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb active"
                >
                  STATE
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  HISTORY
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  PLUGINS
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  SCHEMA
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  STRUCTURE
                </button>
              </li>
               
              <li
                class="s-G9CdWS8GObkb"
              >
                <button
                  class="s-G9CdWS8GObkb"
                >
                  SNAPSHOTS
                </button>
              </li>
            </ul>
            <!--&lt;TabsMenu&gt;-->
             
            <section
              class="split-view s-f-Fay03D4qzv"
            >
              <div
                class="left-panel s-rhUG6P2W5qiw"
                slot="left"
              >
                <div
                  class="top-row s-rhUG6P2W5qiw"
                >
                  <h2
                    class="s-rhUG6P2W5qiw"
                  >
                    Current doc
                  </h2>
                   
                  <button
                    class="s-NMyTiX1zi6rz"
                  >
                    log
                  </button>
                  <!--&lt;Button&gt;-->
                </div>
                 
                <ul
                  class="tree-view svelte-tree-view s-kiBEL6mEiIqN"
                >
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[0]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      type
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="string"
                    >
                      "doc"
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                      <button
                        class="log-copy-button s-oLgcmvmGJcuH"
                      >
                        log
                      </button>
                       
                      <button
                        class="log-copy-button s-oLgcmvmGJcuH"
                      >
                        copy
                      </button>
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-oLgcmvmGJcuH collapsed"
                    data-tree-id="[1]"
                  >
                    <button
                      class="arrow-btn collapsed s-oLgcmvmGJcuH"
                    >
                      â–¶
                    </button>
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH has-children"
                    >
                      content
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH has-children"
                      data-type="array"
                    >
                       [] 1 items
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                      <button
                        class="log-copy-button s-oLgcmvmGJcuH"
                      >
                        log
                      </button>
                       
                      <button
                        class="log-copy-button s-oLgcmvmGJcuH"
                      >
                        copy
                      </button>
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                </ul>
                <!--&lt;TreeView&gt;-->
              </div>
               
              <div
                class="right-panel s-rhUG6P2W5qiw"
                slot="right"
              >
                <div
                  class="top-row s-rhUG6P2W5qiw"
                >
                  <h2
                    class="s-rhUG6P2W5qiw"
                  >
                    Selection
                  </h2>
                   
                  <button
                    class="selection-btn s-NMyTiX1zi6rz"
                  >
                    <span
                      class="caret-icon s-rhUG6P2W5qiw"
                    />
                  </button>
                  <!--&lt;Button&gt;-->
                </div>
                 
                <ul
                  class="tree-view svelte-tree-view s-kiBEL6mEiIqN"
                >
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[0]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      jsonID
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="string"
                    >
                      "text"
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[1]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      empty
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="boolean"
                    >
                      true
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[2]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      anchor
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="number"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[3]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      from
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="number"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[4]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      head
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="number"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                  <li
                    class="row s-oLgcmvmGJcuH"
                    data-tree-id="[5]"
                  >
                     
                    <div
                      class="node-key s-oLgcmvmGJcuH p-left"
                    >
                      to
                      :
                    </div>
                     
                    <div
                      class="node-value s-oLgcmvmGJcuH"
                      data-type="number"
                    >
                      1
                    </div>
                     
                    <div
                      class="buttons s-oLgcmvmGJcuH"
                    >
                       
                    </div>
                  </li>
                   
                  
                  <!--&lt;TreeViewNode&gt;-->
                </ul>
                <!--&lt;TreeView&gt;-->
                 
                <div
                  class="s-rhUG6P2W5qiw"
                >
                  <h2
                    class="s-rhUG6P2W5qiw"
                  >
                    Active marks
                  </h2>
                   
                  <div
                    class="no-marks s-rhUG6P2W5qiw"
                  >
                    No active marks
                  </div>
                </div>
                 
                <div
                  class="s-rhUG6P2W5qiw"
                >
                  <h2
                    class="s-rhUG6P2W5qiw"
                  >
                    Document stats
                  </h2>
                   
                  <ul
                    class="tree-view svelte-tree-view s-kiBEL6mEiIqN"
                  >
                    <li
                      class="row s-oLgcmvmGJcuH"
                      data-tree-id="[0]"
                    >
                       
                      <div
                        class="node-key s-oLgcmvmGJcuH p-left"
                      >
                        nodeSize
                        :
                      </div>
                       
                      <div
                        class="node-value s-oLgcmvmGJcuH"
                        data-type="number"
                      >
                        4
                      </div>
                       
                      <div
                        class="buttons s-oLgcmvmGJcuH"
                      >
                         
                      </div>
                    </li>
                     
                    
                    <!--&lt;TreeViewNode&gt;-->
                    <li
                      class="row s-oLgcmvmGJcuH"
                      data-tree-id="[1]"
                    >
                       
                      <div
                        class="node-key s-oLgcmvmGJcuH p-left"
                      >
                        childCount
                        :
                      </div>
                       
                      <div
                        class="node-value s-oLgcmvmGJcuH"
                        data-type="number"
                      >
                        1
                      </div>
                       
                      <div
                        class="buttons s-oLgcmvmGJcuH"
                      >
                         
                      </div>
                    </li>
                     
                    
                    <!--&lt;TreeViewNode&gt;-->
                  </ul>
                  <!--&lt;TreeView&gt;-->
                </div>
              </div>
            </section>
            <!--&lt;SplitView&gt;-->
            <!--&lt;StateTab&gt;-->
          </div>
        </div>
      </div>
      <!--&lt;FloatingDock&gt;-->
    </section>
    <!--&lt;DevTools&gt;-->
  </div>
</body>
`;
